libraries/src/MVC/Controller/AdminController.php

```
@@ -140,10 +141,14 @@ public function delete()
             $model = $this->getModel();
 
             // Remove the items.
-            if ($model->delete($cid)) {
-                $this->setMessage(Text::plural($this->text_prefix . '_N_ITEMS_DELETED', \count($cid)));
-            } else {
-                $this->setMessage($model->getError(), 'error');
+            try {
+                if ($model->delete($cid)) {
+                    $this->setMessage(Text::plural($this->text_prefix . '_N_ITEMS_DELETED', \count($cid)));
+                } else {
+                    $this->setMessage($model->getError(), 'error');
+                }
+            } catch (ModelExceptionInterface $e) {
```

As the object needs to be an instance of Exception, it would be clearer to have a concrete class here.

---

Interfaces are pretty common for this stuff. For example in the FIG container standards which we use https://github.com/php-fig/container/blob/master/src/ContainerExceptionInterface.php

---

I agree with using an interface instead of a concrete Exception subclass. 

In my own model I want to be able to have different custom exception classes. Some of them may be extending from the general RuntimeException but some of them may be BadMethodCall (e.g. you tried to call an AJAX request in Akeeba Backup for an AJAX method no longer implemented — this would indicate a browser cache problem) or InvalidArgumentException. All of these can implement the ModelExceptionInterface. Not all of these can extend a common exception class, though.

---

Then at least make it extends Throwable, having a plain interface looks odd to me.

---

If it extends Throwable we have the same issue. I have the following exception class:
```php
class DoesNotEatBananas extends \InvalidArgumentException {}
```
If we have a ModelException which extends Throwable how do I refactor this code? I cannot! If I extend from ModelException it is no longer an InvalidArgumentException, it's a generic Throwable. I have to introduce an NicholasInvalidArgumentException extends ModelException, refactor my exceptions to use it and then refactor all of my code which was catching InvalidArgumentException to now catch NicholasInvalidArgumentException. However, this code will not catch real InvalidArgumentException so now I have to do a double catch with code duplication or do some code gymnastics introducing a variable to store a thrown exception and use if-blocks to check its type, completely nullifying the whole idea of exception handling!

Or, we can introduce a ModelExceptionInterface. I change my exception class to read
```php
class DoesNotEatBananas extends \InvalidArgumentException implements ModelExceptionInterface {}
```
and I have to not make any change in my code whatsoever.

Using an interface is the CORRECT way to do it. Doing anything else is not just a massive b/c break, it forces developers to NOT use custom exception types for fear that their code will again break massively when Joomla changes the hierarchy of its exceptions.

Think about it. It's very obvious to me.

---

Does not every PHP Exception implement Throwable?

---

If you change George's code to catch a Throwable instead of an Interface then the code triggers on all exceptions, even those thrown by PHP's ErrorException (an otherwise fatal error) which ends up masking all errors.

George is using an Interface so you, the developer, can control _which_ exceptions have to be caught. You want ErrorException, ParseError etc to bubble up to the main error handler. You don't want to mask them. Basically, you don't want to transform legit bugs which are easy to catch and debug into much more subtle and hard to debug bugs.

---

We have been talking before about the interface and not to catch here Throwable. The `ModelExceptionInterface` should at least extends Throwable. Catching here Throwable makes no sense, especially when we have an interface, that's clear.

---

[You cannot implement Throwable](https://www.php.net/throwable) per the PHP documentation, you can only extend from Exception which wouldn't work for the reason I explained. Unless the PHP documentation is misleading?

---

I'm talking about:

```
interface ModelExceptionInterface extends \Throwable {};
class MyModelexceptionException extends AnyExceptionYouWant implements ModelExceptionInterface {};
```

Otherwise the `ModelExceptionInterface` can be used everywhere and not only in an Exception context, which makes no sense from what I get in this pr.

---

Ah, the documentation was misleading, you can use Throwable in interfaces. Yes, I agree with this.